src/mbgl/renderer/style_diff.cpp

```diff
@@ -0,0 +1,56 @@
+#include <mbgl/renderer/style_diff.hpp>
+#include <mbgl/style/layer_impl.hpp>
+#include <mbgl/util/immutable.hpp>
+#include <mbgl/util/variant.hpp>
+#include <mbgl/util/longest_common_subsequence.hpp>
+
+namespace mbgl {
+
+template <class T>
+StyleDifference<T> diff(const std::vector<T>& a, const std::vector<T>& b) {
+    std::vector<T> lcs;
+
+    auto eq = [] (const T& lhs, const T& rhs) {
+        return std::tie(lhs->id, lhs->type)
+            == std::tie(rhs->id, lhs->type);
+    };
+
+    longest_common_subsequence(a.begin(), a.end(), b.begin(), b.end(), std::back_inserter(lcs), eq);
```

Comment:
This LCS implementation is the unrefined O(ND) space algorithm from [the classic paper](http://xmailserver.org/diff2.pdf). I'm not sure if we need a full LCS, versus a more basic unordered set difference -- it probably depends if we want to implement a "move layer" operation rather than the current "remove and re-add" implementation (see https://github.com/mapbox/mapbox-gl-js/issues/4142).

If we do end up keeping LCS, we should probably implement the O(N) space refinement presented in the paper. I haven't yet invested the time to understand it or write a high-quality implementation.